documentation/man/hmmer.man
.TH "HMMER" 1 "@HMMER_DATE@" "HMMER @HMMER_VERSION@" "HMMER Manual"

.SH NAME

HMMER - profile HMMs for biological sequence analysis

.SH SYNOPSIS

.B hmmalign
  Align sequences to a profile

.B hmmbuild
  Construct profile(s) from multiple sequence alignment(s)

.B hmmconvert
  Convert profile file to various HMMER and non-HMMER formats

.B hmmemit
  Sample sequences from a profile

.B hmmfetch
  Retrieve profile HMM(s) from a file

.B hmmpgmd
  Daemon used for hmmer.org web server, for phmmer-, hmmsearch-, and hmmscan-like search 

.B hmmpress
  Prepare an HMM database for hmmscan

.B hmmscan
  Search protein sequence(s) against a protein profile database

.B hmmsearch
  Search protein profile(s) against a protein sequence database

.B hmmsim
  Collect profile score distributions on random sequences

.B hmmstat
  Summary statistics for a profile file

.B jackhmmer
  Iterative search of a protein sequence against a protein sequence database  

.B nhmmer
  Search nucleotide sequence(s), alignment(s), or profile(s) against a nucleotide sequence database

.B nhmmscan
  Search nucleotide sequence(s) against a nucleotide profile database

.B phmmer
  Search a protein sequence(s) against a protein sequence database


.SH DESCRIPTION

HMMER is a suite of several programs for biological sequence alignment
and database homology search. It uses probabilistic models called
"profile hidden Markov models" (profile HMMs) to represent the likely
evolutionary homologs of a single sequence or a multiple alignment of
a sequence family. A main avenue of research is to improve the
evolutionary predictive models in HMMER to be able to recognize and
accurately align increasingly remote homologs, distant in time.

HMMER is also used as an organizational tool, to group the
exponentially growing number of biological sequences into a vastly
smaller set of well-annotated sequence families. New sequences can be
annotated by comparison against curated sequence family databases of
prebuilt HMMER profiles, in addition or instead of comparison to the
entire sequence database. Databases such as Pfam, SMART, and
TIGRfams, among others, are based on this principle.

HMMER is used in three main modes: to search a sequence database for
new homologs of a sequence or a sequence family; to search a profile
database (like Pfam) to find what known family a query sequence
belongs to, or what domains it has; and to automatically construct
large multiple alignments (i.e. with an effectively unlimited number
of sequences) using a profile representative of a sequence family.


Suppose you have a multiple sequence alignment of a sequence family of
interest, and you want to search a sequence database for additional
homologs. The
.B hmmbuild 
program builds profile(s) from multiple alignment(s). 
The
.B hmmsearch
program searches protein profile(s) against a protein sequence database,
and 
.B nhmmer
searches nucleotide profile(s) against a nucleotide sequence database.

Suppose you have a single sequence of interest, and you want to search
a sequence database for additional homologs. The
.B phmmer
program searches a single protein sequence against a protein sequence
database. The
.B jackhmmer 
program does the same thing but iteratively -- homologs detected in a
previous round are incorporated into a new profile, and the new
profile is searched again. 
.B phmmer 
is used like BLASTP, and 
.B jackhmmer
is used like a protein PSI-BLAST. The
.B nhmmer
program searches a single nucleotide sequence against a nucleotide sequence.

Suppose you have sequence(s) that you want to analyze using a
HMMER-based profile HMM database like Pfam (http://pfam.sanger.ac.uk).
The
.B hmmpress
program formats a profile HMM flatfile (such as the file you
would download from Pfam) into a HMMER binary database.
The 
.B hmmscan
program searches protein sequence(s) against that database.
The 
.B nhmmscan
program can similarly search nucleotide sequence(s) against
a pressed database of nucleotide profiles, such as from 
Dfam (http://dfam.janelia.org).


Suppose you want to align lots of sequences. You can construct a
manageably small alignment of a representative set of sequences,
build a profile with
.BR hmmbuild ,
and use the
.B hmmalign 
program to align any number of sequences to that profile.

HMMER also includes some auxiliary tools for working with large
profile databases.
.B hmmfetch 
fetches one or more profiles from a database.
.B hmmstat 
prints summary statistics about a profile file.

For compatibility with other profile software and previous versions of
HMMER, the
.B hmmconvert
program converts profiles to a few other formats. We intend to add
more support for other formats over time.

The
.B hmmemit 
program generates (simulates) "homologous" sequences by sampling from
a profile. It can also generate a "consensus" sequence.

The 
.B hmmsim
program is a simulator used for collecting statistics about score
distributions on random sequences. 

Each program has its own man page.


.SH SEE ALSO 

This is a summary man page for the entire HMMER3 package.
See individual man pages
.BR [hmmbuild(1) ,
for example]
for usage, options, and description of each program in the package.

.PP
For complete documentation, see the user guide that came with your
HMMER distribution (Userguide.pdf); or see the HMMER web page
(@HMMER_URL@).


.SH COPYRIGHT

.nf
@HMMER_COPYRIGHT@
@HMMER_LICENSE@
.fi

For additional information on copyright and licensing, see the file
called COPYRIGHT in your HMMER source distribution, or see the HMMER
web page 
(@HMMER_URL@).


.SH AUTHOR

.nf
Eddy/Rivas Laboratory
Janelia Farm Research Campus
19700 Helix Drive
Ashburn VA 20147 USA
http://eddylab.org
.fi


src/hmmer.c
/* General routines used throughout HMMER.
 * 
 * Contents:
 *   1. Miscellaneous functions for H3
 *   2. Unit tests
 *   3. Test driver
 *   4. License and copyright 
 *
 * SRE, Fri Jan 12 13:19:38 2007 [Janelia] [Franz Ferdinand, eponymous]
 * SVN $Id$
 */

#include "p7_config.h"

#include <math.h>
#include <float.h>

#include "easel.h"
#include "esl_getopts.h"
#include "hmmer.h"


/*****************************************************************
 * 1. Miscellaneous functions for H3
 *****************************************************************/

/* Function:  p7_banner()
 * Synopsis:  print standard HMMER application output header
 * Incept:    SRE, Wed May 23 10:45:53 2007 [Janelia]
 *
 * Purpose:   Print the standard HMMER command line application banner
 *            to <fp>, constructing it from <progname> (the name of the
 *            program) and a short one-line description <banner>.
 *            For example, 
 *            <p7_banner(stdout, "hmmsim", "collect profile HMM score distributions");>
 *            might result in:
 *            
 *            \begin{cchunk}
 *            # hmmsim :: collect profile HMM score distributions
 *            # HMMER 3.0 (May 2007)
 *            # Copyright (C) 2004-2007 HHMI Janelia Farm Research Campus
 *            # Freely licensed under the Janelia Software License.
 *            # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 *            \end{cchunk}
 *              
 *            <progname> would typically be an application's
 *            <argv[0]>, rather than a fixed string. This allows the
 *            program to be renamed, or called under different names
 *            via symlinks. Any path in the <progname> is discarded;
 *            for instance, if <progname> is "/usr/local/bin/hmmsim",
 *            "hmmsim" is used as the program name.
 *            
 * Note:    
 *    Needs to pick up preprocessor #define's from p7_config.h,
 *    as set by ./configure:
 *            
 *    symbol          example
 *    ------          ----------------
 *    HMMER_VERSION   "3.0"
 *    HMMER_DATE      "May 2007"
 *    HMMER_COPYRIGHT "Copyright (C) 2004-2007 HHMI Janelia Farm Research Campus"
 *    HMMER_LICENSE   "Freely licensed under the Janelia Software License."
 *
 * Returns:   (void)
 */
void
p7_banner(FILE *fp, char *progname, char *banner)
{
  char *appname = NULL;

  if (esl_FileTail(progname, FALSE, &appname) != eslOK) appname = progname;

  fprintf(fp, "# %s :: %s\n", appname, banner);
  fprintf(fp, "# HMMER %s (%s); %s\n", HMMER_VERSION, HMMER_DATE, HMMER_URL);
  fprintf(fp, "# %s\n", HMMER_COPYRIGHT);
  fprintf(fp, "# %s\n", HMMER_LICENSE);
  fprintf(fp, "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n");

  if (appname != NULL) free(appname);
  return;
}


/* Function:  p7_CreateDefaultApp()
 * Synopsis:  Initialize a small/simple/standard HMMER application
 * Incept:    SRE, Thu Oct 28 15:03:21 2010 [Janelia]
 *
 * Purpose:   Identical to <esl_getopts_CreateDefaultApp()>, but 
 *            specialized for HMMER. See documentation in 
 *            <easel/esl_getopts.c>.
 *
 * Args:      options - array of <ESL_OPTIONS> structures for getopts
 *            nargs   - number of cmd line arguments expected (excl. of cmdname)
 *            argc    - <argc> from main()
 *            argv    - <argv> from main()
 *            banner  - optional one-line description of program (or NULL)
 *            usage   - optional one-line usage hint (or NULL)
 *
 * Returns:   ptr to new <ESL_GETOPTS> object.
 * 
 *            On command line errors, this routine prints an error
 *            message to <stderr> then calls <exit(1)> to halt
 *            execution with abnormal (1) status.
 *            
 *            If the standard <-h> option is seen, the routine prints
 *            the help page (using the data in the <options> structure),
 *            then calls <exit(0)> to exit with normal (0) status.
 *            
 * Xref:      J7/3
 * 
 * Note:      The only difference between this and esl_getopts_CreateDefaultApp()
 *            is to call p7_banner() instead of esl_banner(), to get HMMER
 *            versioning info into the header. There ought to be a better way
 *            (perhaps using PACKAGE_* define's instead of HMMER_* vs. EASEL_*
 *            define's in esl_banner(), thus removing the need for p7_banner).
 */
ESL_GETOPTS *
p7_CreateDefaultApp(ESL_OPTIONS *options, int nargs, int argc, char **argv, char *banner, char *usage)
{
  ESL_GETOPTS *go = NULL;

  go = esl_getopts_Create(options);
  if (esl_opt_ProcessCmdline(go, argc, argv) != eslOK ||
      esl_opt_VerifyConfig(go)               != eslOK) 
    {
      printf("Failed to parse command line: %s\n", go->errbuf);
      if (usage != NULL) esl_usage(stdout, argv[0], usage);
      printf("\nTo see more help on available options, do %s -h\n\n", argv[0]);
      exit(1);
    }
  if (esl_opt_GetBoolean(go, "-h") == TRUE) 
    {
      if (banner != NULL) p7_banner(stdout, argv[0], banner);
      if (usage  != NULL) esl_usage (stdout, argv[0], usage);
      puts("\nOptions:");
      esl_opt_DisplayHelp(stdout, go, 0, 2, 80);
      exit(0);
    }
  if (esl_opt_ArgNumber(go) != nargs) 
    {
      puts("Incorrect number of command line arguments.");
      esl_usage(stdout, argv[0], usage);
      printf("\nTo see more help on available options, do %s -h\n\n", argv[0]);
      exit(1);
    }
  return go;
}


/* Function:  p7_AminoFrequencies()
 * Incept:    SRE, Fri Jan 12 13:46:41 2007 [Janelia]
 *
 * Purpose:   Fills a vector <f> with amino acid background frequencies,
 *            in [A..Y] alphabetic order, same order that Easel digital
 *            alphabet uses. Caller must provide <f> allocated for at
 *            least 20 floats.
 *            
 *            These were updated 4 Sept 2007, from Swiss-Prot 50.8,
 *            (Oct 2006), counting over 85956127 (86.0M) residues.
 *
 * Returns:   <eslOK> on success.
 */
int
p7_AminoFrequencies(float *f)
{
  f[0] = 0.0787945;		/* A */
  f[1] = 0.0151600;		/* C */
  f[2] = 0.0535222;		/* D */
  f[3] = 0.0668298;		/* E */
  f[4] = 0.0397062;		/* F */
  f[5] = 0.0695071;		/* G */
  f[6] = 0.0229198;		/* H */
  f[7] = 0.0590092;		/* I */
  f[8] = 0.0594422;		/* K */
  f[9] = 0.0963728;		/* L */
  f[10]= 0.0237718;		/* M */
  f[11]= 0.0414386;		/* N */
  f[12]= 0.0482904;		/* P */
  f[13]= 0.0395639;		/* Q */
  f[14]= 0.0540978;		/* R */
  f[15]= 0.0683364;		/* S */
  f[16]= 0.0540687;		/* T */
  f[17]= 0.0673417;		/* V */
  f[18]= 0.0114135;		/* W */
  f[19]= 0.0304133;		/* Y */
  return eslOK;
}

/*****************************************************************
 * 2. Unit tests
 *****************************************************************/
#ifdef p7HMMER_TESTDRIVE

static void
utest_alphabet_config(int alphatype)
{
  char         *msg = "HMMER alphabet config unit test failed";
  ESL_ALPHABET *abc = NULL;

  if ((abc = esl_alphabet_Create(alphatype)) == NULL) esl_fatal(msg);
  if (abc->K  > p7_MAXABET)                           esl_fatal(msg);
  if (abc->Kp > p7_MAXCODE)                           esl_fatal(msg);
  esl_alphabet_Destroy(abc);
}
#endif /*p7HMMER_TESTDRIVE*/

  

/*****************************************************************
 * 3. Test driver
 *****************************************************************/
#ifdef p7HMMER_TESTDRIVE

/* gcc -o hmmer_utest -g -Wall -I../easel -L../easel -I. -L. -Dp7HMMER_TESTDRIVE hmmer.c -lhmmer -leasel -lm
 * ./hmmer_utest
 */
#include "esl_getopts.h"

static ESL_OPTIONS options[] = {
  /* name           type      default  env  range toggles reqs incomp  help                                       docgroup*/
  { "-h",        eslARG_NONE,   FALSE, NULL, NULL, NULL, NULL, NULL, "show brief help on version and usage",              0 },
  {  0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
};
static char usage[]  = "[-options]";
static char banner[] = "test driver for hmmer.c";

int
main(int argc, char **argv)
{
  ESL_GETOPTS *go = p7_CreateDefaultApp(options, 0, argc, argv, banner, usage);

  utest_alphabet_config(eslAMINO);
  utest_alphabet_config(eslDNA);
  utest_alphabet_config(eslRNA);
  utest_alphabet_config(eslCOINS);
  utest_alphabet_config(eslDICE);

  esl_getopts_Destroy(go);
  return 0;
}
#endif /*p7HMMER_TESTDRIVE*/



/*****************************************************************
 * @LICENSE@
 *****************************************************************/
